### ISSUE TITLE
Don't impose horrible algorithmic burdens on implementors

### ISSUE THREAD

**Issue by hcayless** [TEIC/TEI#1497]
In rather stark contrast to the note in #1496, which says implementors should do whatever they like when a list of models don't have an obvious first choice, a following note says:

> If there are two or more model elements provided for an elementSpec but they have different predicate attributes then the implementation should choose the model whose predicate provides the most specific context (where specific is understood in the same way as in XSLT) 

It's very unkind to ask implementors to use the complex template matching algorithm from XSLT when we refuse to help them at all when there are no `@predicate`s. I'd recommend that if two models match an element then the first matching model should be chosen.

**Comment by lb42** [comment]
What does "first" mean here? First in sequence within the elementSpec? What if there are two elementSpecs, each with one model? Does the first in sequence within the ODD win? We don't  make choices based on sequence anywhere else in ODD, sfaicr

**Comment by hcayless** [comment]
Do you mean two elementSpecs for the same element? Does that ever happen?

**Comment by hcayless** [comment]
AFAICT, there's never going to be a conflict between models in different elementSpecs, because they shouldn't get triggered in conflict. The model for `<ab>` would be triggered when you hit an `<ab>` and the model for `<p>` when you hit a `<p>`. My argument is that if you have more than one model in an elementSpec (or a modelGrp), and you haven't been careful about which one gets priority, then an implementation should just pick the first one that matches. 

I think it's very bad practice not to tell implementors how to implement your spec.

**Comment by lb42** [comment]
Two elementSpecs for the same element happens *all the time* in an ODD, which is where this is supposed to be used. Suppose that, through some outburst of collective insanity, the Council were to start storing  processing models as part of the P5 specs. In my ODD I add a new processing model for the &lt;ab> model. Which one wins?

**Comment by hcayless** [comment]
Ah, I see what you're saying. It would depend on the @mode, wouldn't it? The ODD's going to have to be compiled before it's processed.

**Comment by lb42** [comment]
... which in our case we have not got.

**Comment by hcayless** [comment]
I'd say that's a bug too.

**Comment by jamescummings** [comment]
I believe the reluctance to tell implementors how to do this was based on different forms of implementation having different challenges. In XSLT implementation they can do most-specific-predicate easily enough. In python it would be easier to take 'first matching in order found in the compiled odd' or something.  Either the TEI stays agnostic about this and says "Do what you want be be consistent and document it" or we need to revist this decision.

**Comment by hcayless** [comment]
What I'd prefer to avoid is the possibility that different implementations behave completely differently. It's easy enough to implement the "first matching" algortihm in XSLT—just put them in an xsl:choice, for example. It's not so easy to do that in other languages. And anyway, the spec *does* tell implementors what to do here. It says "do it the hard way".

**Comment by hcayless** [comment]
Re @lb42's point above:
> Two elementSpecs for the same element happens all the time in an ODD, which is where this is supposed to be used. Suppose that, through some outburst of collective insanity, the Council were to start storing processing models as part of the P5 specs. In my ODD I add a new processing model for the <ab> model. Which one wins?

As @jamescummings pointed out in #1498 (and I think he's right), if you define a new model in an ODD, it's going to override any existing models, because models aren't [combinable](http://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-att.combinable.html). So I don't think there's ever going to be ambiguity about model order.

**Comment by tuurma** [comment]
I support the view that first matching model should win. I remember loosing an argument about this at some point, but the key argument was what is to happen when chaining ODDs and combining models. Since discussion in #1498 basically makes it a moot point, I'm happy to return to my original position :-)

Let me list my main arguments:
- algorithm for selecting matching model should be clear, and simple enough to understand for PM's intended users (ie editors as well as developers), preferably easy to specify and document as well
- it should be efficient (and preferably simple) to implement in a range of programming languages

While sticking to XSLT-like matching when building an XSLT implementation of PM is easy enough, it is not the case when doing it in Java, Python or XQuery and introduces significant performance issues there. 

The advantage of doing so is not clear: sure, a group of developers/editors already familiar with XSLT does not need further documentation except saying 'please consult XSLT docs' but for all others it raises the bar for entry quite a bit. It also introduces certain ambiguity in some cases as to which model will be chosen, since template priorities in XSLT can be a hidden killer.

As a side note, the 'first one that matches' approach was deliberately chosen when we started the PM prototype (although I don't remember having very heated discussions there) and sustained a few months later when Sebastian was looking for alternative implementations and exchanged some correspondence with potential implementors on the subject.

**Comment by hcayless** [comment]
Unsurprisingly, I agree with @tuurma :-) (and thanks for the background). It sounds to me like there **was** some sort of inchoate plan for implementing model combination, in which case you would need to work out either precisely how that combination worked in order to sustain a "first matching" rule, or have a more sophisticated algorithm for choosing models (à la XSLT). But in what we have, it wasn't implemented, so models aren't combinable, and there's no reason not to do the simplest thing and choose the first matching option.

**Comment by lb42** [comment]
"inchoate plan" sounds about right for rather a lot of the processing model implementation, I fear. I don't mind how we improve it: I just reserve the right to say "told you so" when we or our successors get round to making models combinable.

**Comment by hcayless** [comment]
If/when we or they do, there will have to be rules about how models combine that are more specific than the rules for other combinable objects. Until the day when the PM is successful enough that people are clamoring for this feature though, I invoke [YAGNI](http://martinfowler.com/bliki/Yagni.html).

**Comment by tuurma** [comment]
Yes, I support that point of view. This is all fairly liquid now and perhaps waiting to see what the approaches towards chaining ODDs and models are in practice before deciding is wisest (and easiest) course for now.

**Comment by tuurma** [comment]
To clarify my comment above: waiting before prescribing a way to combine models. As for now, I propose to remove 'needs discussion' label and formulate 'first one matching in document order' rule.

**Comment by jamescummings** [comment]
I would add to the 'first one matching in ODD document order' rule the assumption for chained ODDs that any imported models are assumed to come _after_ the models in the ODD that imports them. (i.e. if your ODD is pointing to another ODD as a source, and you write new models, those are assumed to come first)

**Comment by tuurma** [comment]
related discussion in #1494

**Comment by tuurma** [comment]
Pasting @hcayless's proposed wording from #1494 

>A "matching" model is one where the element to be processed satisfies the XPath in the @predicate attribute of the model or modelSequence and the current output method matches the method specified in the @output attribute of the model, modelSequence, or a containing modelGrp. A model or modelSequence without a @predicate always matches the element to be processed. A model, modelGrp, or modelSequence without an @output attribute matches any output method.

>Processing Model implementations must execute only the first matching model or modelSequence in document order.

**Comment by hcayless** [comment]
I went ahead and folded my proposed wording into TD-DocumentationElements.

**Comment by tuurma** [comment]
somehow the contradictory wording survived unnoticed till now, PR imminent
